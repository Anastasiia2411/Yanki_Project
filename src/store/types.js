export const PRODUCTS_LOAD = "PRODUCTS_LOAD";
export const ADD_TO_BUY_ARR = "ADD_TO_BUY_ARR";
export const SAVE_BUY_ARR_TO_LOCALE_STORAGE = "SAVE_BUY_ARR_TO_LOCAL_STORAGE";
export const REMOVE_FROM_BUY_ARR = "REMOVE_FROM_BUY_ARR";
export const ADD_TO_LIKE_ARR = "ADD_TO_LIKE_ARR";
export const SAVE_LIKE_ARR_TO_LOCALE_STORAGE = "SAVE_LIKE_ARR_TO_LOCALE_STORAGE";
export const AMOUNT_OF_BUY_PRODUCT = "AMOUNT_OF_BUY_PRODUCT";
export const PURCHASES_MODAL_OPEN = "PURCHASES_MODAL_OPEN";
export const BASKET_MODAL_OPEN = "PURCHASES_MODAL_OPEN";
export const LIKED_MODAL_OPEN = "LIKED_MODAL_OPEN";
export const PRODUCTS_LOAD_FAILED = "PRODUCTS_LOAD_FAILED";
export const FORM_MODAL_OPEN = "FORM_MODAL_OPEN";
export const CLOSE_MODAL_FORM = "CLOSE_MODAL_FORM";
export const MODAL_FORM_SEND_DATA_ABOUT_USERS = "MODAL_FORM_SEND_DATA_ABOUT_USERS";
export const MODAL_FORM_SEND_DATA_FROM_BUY_PAGE = "MODAL_FORM_SEND_DATA_FROM_BUY_PAGE";
export const DELETE_PURCHASE_FROM_BUY_PAGE = "DELETE_PURCHASE_FROM_BUY_PAGE";
export const CLOSE_MODAL_WHEN_PURCHASES_ARE_EMPTY="CLOSE_MODAL_WHEN_PURCHASES_ARE_EMPTY"
export const OPEN_MODAL_WHEN_PURCHASES_ARE_EMPTY ="OPEN_MODAL_WHEN_PURCHASES_ARE_EMPTY"
