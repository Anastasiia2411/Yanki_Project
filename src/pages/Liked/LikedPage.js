import React, { useContext } from "react";
import { LikedHeader, LikedParagraph, ParagraphWrapper } from "./LikedPage.style";
import { useDispatch, useSelector, shallowEqual } from "react-redux";

import { ProductElement } from "../../components/Products/Product";
import { Modal } from "../../components/Modal/Modal";
import { ButtonsWrapper } from "../../components/Modal/Modal.style";
import { ButtonComponent } from "../../components/Modal/Button";
import {
    addProductToPurchases,
    addToLikedProducts,
    addProductToBasketFromLikedProductsConfirmModal
} from "../../store/actions";
import { ViewContext } from "../../components/Layout/Layout";
import { Table, Th, Tr, TableWrapper } from "../../components/Products/Products.style";
import { BuysParagraph, BuysParagraphWrapper } from "../Buys/BuysPage.style";

const color = "#E0BEA2";

export function LikedPage() {
    const dispatch = useDispatch();
    const likedProducts = useSelector(state => state.like.likedProducts, shallowEqual);
    const likedProductConfirmModalItem = useSelector(state => state.likedPageAddProductToBasket.likedProductConfirmModalItem, shallowEqual);
    const { view } = useContext(ViewContext);

    function addToPurchasesArr(item) {
        dispatch(addProductToPurchases(item));
    }

    function addToLikedProductsArr(item) {
        dispatch(addToLikedProducts(item));
    }

    function addProductToBasketFromLikedProductsConfirmModalDispatch(item) {

        console.log(likedProductConfirmModalItem);
        dispatch(addProductToBasketFromLikedProductsConfirmModal(item));
    }

    const addLikedProductToPurchasesAndCloseConfirmModal = () => {
        addToPurchasesArr(likedProductConfirmModalItem);
        addProductToBasketFromLikedProductsConfirmModalDispatch(null);
    };

    const closeLikedProductConfirmModalWithoutAddingProduct = () => {
        addProductToBasketFromLikedProductsConfirmModalDispatch(null);
    };

    return (
        <>
            <LikedHeader>Вподобане:</LikedHeader>
            {likedProducts.length > 0 ? (
                view === "list" ? (
                <TableWrapper>
                    <Table>
                        <tbody>
                        <Tr>
                            <Th>Фото</Th>
                            <Th>Назва</Th>
                            <Th>Ціна</Th>
                            <Th>Артикль</Th>
                            <Th>Колір</Th>
                            <Th>Кнопка</Th>
                        </Tr>
                        {likedProducts.map((product) => {
                            return (
                                <ProductElement
                                    view={view}
                                    product={product}
                                    key={product.article}
                                    buttonText={"Додати в корзину"}
                                    isLiked={likedProducts.some(
                                        (likedItem) => likedItem.article === product.article
                                    )}
                                    onAddToLikedButtonClick={addToLikedProductsArr}
                                    handleButtonClick={
                                        addProductToBasketFromLikedProductsConfirmModalDispatch
                                    }
                                />
                            );
                        })}
                        </tbody>
                    </Table>
                </TableWrapper>
                ) : (
                    <ParagraphWrapper view={view}>
                        {likedProducts.length > 0 ? (
                            likedProducts.map((product) => {
                                return (
                                    <ProductElement
                                        view={view}
                                        product={product}
                                        key={product.article}
                                        buttonText={"Додати в корзину"}
                                        isLiked={likedProducts.some(
                                            (likedItem) => likedItem.article === product.article
                                        )}
                                        onAddToLikedButtonClick={addToLikedProductsArr}
                                        handleButtonClick={
                                            addProductToBasketFromLikedProductsConfirmModalDispatch
                                        }
                                    />
                                );
                            })
                        ) : null}
                    </ParagraphWrapper>
                )) : (
                <BuysParagraphWrapper>
                    <BuysParagraph>Немає товарів до покупки</BuysParagraph>
                </BuysParagraphWrapper>
            )
            }
            {likedProductConfirmModalItem && (
                <Modal
                    headerBackgroundColor={color}
                    header="Ви хочете додати цей товар до Вашого кошика?"
                    closeButton={false}
                    text="Ви зможете переглянути всі вибрані товари у вкладенні 'Кошик'"
                    onClose={closeLikedProductConfirmModalWithoutAddingProduct}
                >
                    <ButtonsWrapper>
                        <ButtonComponent
                            text="Так"
                            handleClick={addLikedProductToPurchasesAndCloseConfirmModal}
                            background={color}
                        />
                        <ButtonComponent
                            text="Ні"
                            handleClick={closeLikedProductConfirmModalWithoutAddingProduct}
                            background={color}
                        />
                    </ButtonsWrapper>
                </Modal>
            )}
        </>
    );
}