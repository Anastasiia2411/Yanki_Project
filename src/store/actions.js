import {
    PRODUCTS_LOAD,
    ADD_TO_BUY_ARR,
    SAVE_BUY_ARR_TO_LOCALE_STORAGE,
    REMOVE_FROM_BUY_ARR,
    ADD_TO_LIKE_ARR,
    SAVE_LIKE_ARR_TO_LOCALE_STORAGE,
    AMOUNT_OF_BUY_PRODUCT,
    PURCHASES_MODAL_OPEN,
    BASKET_MODAL_OPEN,
    LIKED_MODAL_OPEN,
    PRODUCTS_LOAD_FAILED,
    FORM_MODAL_OPEN,
    CLOSE_MODAL_FORM,
    MODAL_FORM_SEND_DATA_ABOUT_USERS,
    MODAL_FORM_SEND_DATA_FROM_BUY_PAGE,
    DELETE_PURCHASE_FROM_BUY_PAGE,
    OPEN_MODAL_WHEN_PURCHASES_ARE_EMPTY,
    CLOSE_MODAL_WHEN_PURCHASES_ARE_EMPTY,
} from "./types";

export function getAllProducts() {
    return async dispatch => {
        try {
            const informResponse = await fetch("./inform.json");
            const products = await informResponse.json();
            dispatch({
                type: PRODUCTS_LOAD,
                data: products,
            });
        } catch (error) {
            console.error("Error fetching products:", error);
            dispatch({
                type: PRODUCTS_LOAD_FAILED,
                error: error.message,
            });
        }
    };
}

export function addProductToPurchases(item) {
    if (!item || typeof item !== "object" || !item.name || !item.photo || !item.article) {
        console.error("Invalid item object:", item);
        return null;
    }
    return { type: ADD_TO_BUY_ARR, item };
}

export function BuyConfirmModalOpen(item) {
    return { type: PURCHASES_MODAL_OPEN, item };
}

export function deleteConfirmModal(item) {
    return { type: BASKET_MODAL_OPEN, item };
}

export function addProductToBasketFromLikedProductsConfirmModal(item) {
    return { type: LIKED_MODAL_OPEN, item };
}

export function addToLikedProducts(item) {
    if (!item || typeof item !== "object" || !item.name || !item.photo || !item.article) {
        console.error("Invalid item object:", item);
        return null;
    }
    return { type: ADD_TO_LIKE_ARR, item };
}

export function removeProductFromPurchases(item) {
    return { type: REMOVE_FROM_BUY_ARR, item };
}

export const savePurchasesArrToLocalStorage = () => {
    return (dispatch, getState) => {
        const purchases = getState().buy.purchases;
        localStorage.setItem("purchases", JSON.stringify(purchases));
        dispatch({ type: SAVE_BUY_ARR_TO_LOCALE_STORAGE });
    };
};

export const saveLikedProductsArrToLocalStorage = () => {
    return (dispatch, getState) => {
        const likedProducts = getState().like.likedProducts;
        localStorage.setItem("likedProducts", JSON.stringify(likedProducts));
        dispatch({ type: SAVE_LIKE_ARR_TO_LOCALE_STORAGE });
    };
};

export const amountOfBuyProduct = () => {
    return { type: AMOUNT_OF_BUY_PRODUCT };
};

export function OpenFormModal() {
    return { type: FORM_MODAL_OPEN };
}

export function CloseFormModal() {
    return { type: CLOSE_MODAL_FORM };
}

export function sendDataAboutUsers(usersInformObj) {
   return { type: MODAL_FORM_SEND_DATA_ABOUT_USERS, usersInformObj};

}

export function sendDataAboutBuyProduct(usersBuyProductsArr) {
    return { type: MODAL_FORM_SEND_DATA_FROM_BUY_PAGE, usersBuyProductsArr};
}

export function clearTheArrayOfBuyProducts(){
    return { type: DELETE_PURCHASE_FROM_BUY_PAGE};
}

export function OpenModalWhichSaysAboutOfAbsenceProduct(){
    return { type: OPEN_MODAL_WHEN_PURCHASES_ARE_EMPTY};
}

export function CloseModalWhichSaysAboutOfAbsenceProduct(){
    return { type: CLOSE_MODAL_WHEN_PURCHASES_ARE_EMPTY};
}