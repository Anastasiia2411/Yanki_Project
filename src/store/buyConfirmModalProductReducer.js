import { PURCHASES_MODAL_OPEN } from "./types";

const defaultState = {
    addedCartItem: null,
};

export default function buyConfirmModalProductReducer(state = defaultState, action) {
    switch (action.type) {
        case PURCHASES_MODAL_OPEN:
            return {...state, addedCartItem: action.item}
        default:
            return state;
    }
}