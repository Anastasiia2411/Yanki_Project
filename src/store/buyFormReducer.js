import { FORM_MODAL_OPEN, CLOSE_MODAL_FORM, MODAL_FORM_SEND_DATA_FROM_BUY_PAGE, MODAL_FORM_SEND_DATA_ABOUT_USERS, OPEN_MODAL_WHEN_PURCHASES_ARE_EMPTY, CLOSE_MODAL_WHEN_PURCHASES_ARE_EMPTY} from "./types";

const defaultState = {
    openForm: false,
    sendDataFromBuyPage: [],
    sendDataAboutUsers: {},
    purchasesAreEmpty:false,
};

export default function buyFormReducer(state = defaultState, action) {
    switch (action.type) {
        case FORM_MODAL_OPEN:
            return { ...state, openForm: true };
        case CLOSE_MODAL_FORM:
            return { ...state, openForm: false };
        case MODAL_FORM_SEND_DATA_FROM_BUY_PAGE:
            return { ...state, sendDataFromBuyPage: [...action.usersBuyProductsArr] };
        case MODAL_FORM_SEND_DATA_ABOUT_USERS:
            return { ...state, sendDataAboutUsers: action.usersInformObj };
        case OPEN_MODAL_WHEN_PURCHASES_ARE_EMPTY:
            return {...state, purchasesAreEmpty: true}
        case CLOSE_MODAL_WHEN_PURCHASES_ARE_EMPTY:
            return {...state, purchasesAreEmpty: false}
        default:
            return state;
    }
}