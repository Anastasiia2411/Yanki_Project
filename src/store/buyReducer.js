import {
    ADD_TO_BUY_ARR,
    AMOUNT_OF_BUY_PRODUCT,
    REMOVE_FROM_BUY_ARR,
    SAVE_BUY_ARR_TO_LOCALE_STORAGE,
    DELETE_PURCHASE_FROM_BUY_PAGE
} from "./types";

const localStorageBuyArr = JSON.parse(localStorage.getItem("purchases"));

const defaultState = {
    purchases: localStorageBuyArr === null ? [] : localStorageBuyArr,
    amountOfBuyProducts: 0
};

export default function buyReducer(state = defaultState, action) {
    switch (action.type) {
        case ADD_TO_BUY_ARR:
            const itemBuyExists = state.purchases.some((obj) => {
                return obj.article === action.item.article;
            });
            if (itemBuyExists) {
                return {
                    ...state,
                    purchases: state.purchases.map((product) => {
                        if (product.article !== action.item.article) {
                            return product;
                        }
                        return {
                            ...product,
                            count: product.count + 1,
                        };
                    }),
                    amountOfBuyProducts: state.amountOfBuyProducts + 1
                };
            } else {
                return {
                    ...state,
                    purchases: [...state.purchases, { ...action.item, count: 1 }],
                    amountOfBuyProducts: state.amountOfBuyProducts + 1
                };
            }
        case REMOVE_FROM_BUY_ARR:
            const itemWhichIs = state.purchases.some((obj) => {
                return obj.article === action.item.article;
            });
            if (!itemWhichIs) {
                return state;
            } else {
                if (action.item.count === 1) {
                    return {
                        ...state,
                        purchases: state.purchases.filter((el) => el.article !== action.item.article),
                        amountOfBuyProducts: state.amountOfBuyProducts - 1
                    };
                } else if (action.item.count > 1) {
                    return {
                        ...state,
                        purchases: state.purchases.map((el) => {
                            if (el.article !== action.item.article) {
                                return el;
                            }
                            return {
                                ...el,
                                count: el.count - 1,
                            };
                        }),
                        amountOfBuyProducts: state.amountOfBuyProducts - 1
                    };
                }
            }
            return state;
        case SAVE_BUY_ARR_TO_LOCALE_STORAGE:
            return state;
        case AMOUNT_OF_BUY_PRODUCT:
            return { ...state, amountOfBuyProducts: state.purchases.reduce((acc, product) => acc + product.count, 0) };
        case DELETE_PURCHASE_FROM_BUY_PAGE:
            return { ...state, purchases: [] };
        default:
            return state;
    }
}