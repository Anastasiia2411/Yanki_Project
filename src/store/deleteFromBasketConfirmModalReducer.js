import { BASKET_MODAL_OPEN } from "./types";

const defaultState = {
    removedCartItem: null,
};

export default function deleteFromBasketConfirmModalReducer(state = defaultState, action) {
    switch (action.type) {
        case BASKET_MODAL_OPEN:
            return {...state, removedCartItem: action.item}
        default:
            return state;
    }
}