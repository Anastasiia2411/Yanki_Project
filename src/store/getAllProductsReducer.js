import { PRODUCTS_LOAD, PRODUCTS_LOAD_FAILED } from "./types";

const defaultState = {
    allProducts: [],
};

export default function getAllProducts(state = defaultState, action) {
    switch (action.type) {
        case PRODUCTS_LOAD:
            return {
                ...state,
                allProducts: [...state.allProducts, ...action.data],
                isLoading: false,
                error: null,
            };
        case PRODUCTS_LOAD_FAILED:
            return {
                ...state,
                isLoading: false,
                error: action.error,
            };
        default:
            return state;
    }
}
