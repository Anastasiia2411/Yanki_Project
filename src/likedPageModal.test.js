import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";

import { LikedPage } from "./pages/Liked/LikedPage";
import { LIKED_MODAL_OPEN } from "./store/types";

describe("LikedPage", () => {
    const mockStore = configureStore();
    const store = mockStore({
        likedPageAddProductToBasket: {
            likedProductConfirmModalItem: null,
        },
        allProducts: {
            allProducts: [
                {
                    name: "Коричнева куртка",
                    price: "3150 грн",
                    photo: "https://static2.issaplus.com/wa-data/public/photos/96/45/4596/4596.970.jpg",
                    article: "209385",
                    color: "brown"
                },
            ],
        },
        like: {
            likedProducts: [
                {
                    name: "Коричнева куртка",
                    price: "3150 грн",
                    photo: "https://static2.issaplus.com/wa-data/public/photos/96/45/4596/4596.970.jpg",
                    article: "209385",
                    color: "brown"
                },
            ]
        }
    });

    it("opens the modal window with correct text when addToConfirmModal function is called", () => {
        render(
            <Provider store={store}>
                <LikedPage/>
            </Provider>
        );

        const addToCartButton = screen.getByText("Додати в корзину");
        fireEvent.click(addToCartButton);

        const actions = store.getActions().at(-1);

        expect(actions).toStrictEqual({
            type: LIKED_MODAL_OPEN,
            item: {
                name: "Коричнева куртка",
                price: "3150 грн",
                photo: "https://static2.issaplus.com/wa-data/public/photos/96/45/4596/4596.970.jpg",
                article: "209385",
                color: "brown"
            },
        });
    });

    it("matches the snapshot", () => {
        const { asFragment } = render(
            <Provider store={store}>
                <LikedPage/>
            </Provider>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});

describe("LikedPage modal test", () => {
    const mockStore = configureStore();
    const store = mockStore({
        likedPageAddProductToBasket: {
            likedProductConfirmModalItem: {
                name: "Коричнева куртка",
                price: "3150 грн",
                photo: "https://static2.issaplus.com/wa-data/public/photos/96/45/4596/4596.970.jpg",
                article: "209385",
                color: "brown"
            },
        },
        allProducts: {
            allProducts: [
                {
                    name: "Коричнева куртка",
                    price: "3150 грн",
                    photo: "https://static2.issaplus.com/wa-data/public/photos/96/45/4596/4596.970.jpg",
                    article: "209385",
                    color: "brown"
                },
            ],
        },
        like: {
            likedProducts: [
                {
                    name: "Коричнева куртка",
                    price: "3150 грн",
                    photo: "https://static2.issaplus.com/wa-data/public/photos/96/45/4596/4596.970.jpg",
                    article: "209385",
                    color: "brown"
                },
            ]
        }
    });

    it("close  liked modal when users has clicked 'Ні'", () => {
        render(
            <Provider store={store}>
                <LikedPage/>
            </Provider>
        );

        const addToCartButton = screen.getByText("Ні");
        fireEvent.click(addToCartButton);

        const actions = store.getActions().at(-1);

        expect(actions).toStrictEqual({
            type: LIKED_MODAL_OPEN,
            item: null,
        });
    });

    it("close  liked modal when users has clicked 'Так'", () => {
        render(
            <Provider store={store}>
                <LikedPage/>
            </Provider>
        );

        const addToCartButton = screen.getByText("Так");
        fireEvent.click(addToCartButton);

        const actions = store.getActions().at(-1);

        expect(actions).toStrictEqual({
            type: LIKED_MODAL_OPEN,
            item: null,
        });
    });
});
