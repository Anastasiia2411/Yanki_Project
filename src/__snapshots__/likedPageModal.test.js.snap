// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LikedPage matches the snapshot 1`] = `
<DocumentFragment>
  <h2
    class="sc-bcXHqe uqvMk"
  >
    Вподобане:
  </h2>
  <div
    class="sc-dkrFOg klICsf"
  >
    <div
      class="sc-ftTHYK yrluS"
    >
      <img
        alt="jacket"
        height="350"
        src="https://static2.issaplus.com/wa-data/public/photos/96/45/4596/4596.970.jpg"
      />
      <div>
        <svg
          fill="#FFFFFF"
          stroke="red"
          viewBox="0 0 14 13"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.93804 1.15962C8.47954 -0.224381 10.8617 -0.178444 12.3467 1.30925C13.8312 2.79759 13.8823 5.16792 12.5016 6.71402L6.93672 12.2868L1.37314 6.71402C-0.00758263 5.16792 0.0442603 2.79366 1.52801 1.30925C3.01439 -0.176476 5.39194 -0.22635 6.93804 1.15962ZM11.4175 2.23651C10.4332 1.25084 8.84506 1.21081 7.81477 2.13611L6.93869 2.92228L6.06196 2.13676C5.02839 1.21015 3.44357 1.25084 2.45659 2.23782C1.4788 3.21562 1.42958 4.78074 2.33059 5.81497L6.93738 10.429L11.5442 5.81563C12.4458 4.78074 12.3966 3.21759 11.4175 2.23651Z"
            fill="white"
          />
        </svg>
      </div>
      <ul
        class="sc-pyfCe hXQHJ"
      >
        <li>
          Коричнева куртка
        </li>
        <li>
          3150 грн
        </li>
        <li>
          article:209385
        </li>
        <li>
          <button
            class="sc-ipEyDJ ipAkdK"
            data-testid="delete_button"
          >
            Додати в корзину
          </button>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
